lib/Dialect/AIG/Analysis/LongestPathAnalysis.cpp

```diff
@@ -172,47 +262,1100 @@ struct llvm::DenseMapInfo<Object> {
+  // If the result is not cached, compute it and cache it.
+  FailureOr<ArrayRef<OpenPath>> getOrComputeResults(Value value, size_t bitPos);
+
+  // Get the longest paths for the given value and bit position. This returns
+  // an empty array if not pre-computed.
+  ArrayRef<OpenPath> getResults(Value value, size_t bitPos) const;
+
+  // A map from the object to the maximum distance and history.
+  using ObjectToMaxDistance =
+      llvm::MapVector<Object,
+                      std::pair<int64_t, llvm::ImmutableList<DebugPoint>>>;
+
+  void getClosedPaths(SmallVectorImpl<DataflowPath> &results) const;
+  void setTopLevel() { this->topLevel = true; }
+  bool isTopLevel() const { return topLevel; }
+  hw::HWModuleOp getHWModuleOp() const { return module; }
+
+  const auto &getFromInputPortToFanOut() const { return fromInputPortToFanOut; }
+  const auto &getFromOutputPortToFanIn() const { return fromOutputPortToFanIn; }
+  const auto &getFanOutResults() const { return fanOutResults; }
+
+private:
+  void putUnclosedResult(const Object &object, int64_t delay,
+                         llvm::ImmutableList<DebugPoint> history,
+                         ObjectToMaxDistance &objectToMaxDistance);
+
+  // A map from the input port to the farthest fanOut.
+  llvm::MapVector<std::pair<BlockArgument, size_t>, ObjectToMaxDistance>
+      fromInputPortToFanOut;
+
+  // A map from the output port to the farthest fanIn.
+  llvm::MapVector<std::tuple<size_t, size_t>, ObjectToMaxDistance>
+      fromOutputPortToFanIn;
+
+  LogicalResult initializeAndRun(hw::InstanceOp instance);
+  LogicalResult initializeAndRun(hw::OutputOp output);
+
+  // --------------------------------------------------------------------------
+  // Visitors
+  // --------------------------------------------------------------------------
+  LogicalResult visitValue(Value value, size_t bitPos,
+                           SmallVectorImpl<OpenPath> &results);
+  // Module boundary.
+  LogicalResult visit(mlir::BlockArgument argument, size_t bitPos,
+                      SmallVectorImpl<OpenPath> &results);
+  LogicalResult visit(hw::InstanceOp op, size_t bitPos, size_t resultNum,
+                      SmallVectorImpl<OpenPath> &results);
+
+  // Data-movement ops.
+  LogicalResult visit(hw::WireOp op, size_t bitPos,
+                      SmallVectorImpl<OpenPath> &results);
+  LogicalResult visit(comb::ConcatOp op, size_t bitPos,
+                      SmallVectorImpl<OpenPath> &results);
+  LogicalResult visit(comb::ExtractOp op, size_t bitPos,
+                      SmallVectorImpl<OpenPath> &results);
+  LogicalResult visit(comb::ReplicateOp op, size_t bitPos,
+                      SmallVectorImpl<OpenPath> &results);
+  // Track equivalence classes for data movement ops
+  // Treat output as equivalent to input for pure data movement operations
+  llvm::EquivalenceClasses<std::pair<Value, size_t>> ec;
+  DenseMap<std::pair<Value, size_t>, std::pair<Value, size_t>> ecMap;
```

Comment:
Nice use of `EquivalenceClasses`! :partying_face: I'm wondering if you still need to track that separate `ecMap` :thinking: It looks like you use it to map from an entry in the class to the leader, which I think the `ec` can do for you directly :thinking: Might be mistaken though.

Reply:
Yeah totally!